### PR TITLE
Add groups at login if provided in the idToken as an array of strings

### PR DESCRIPTION
**Pull request by DriesVerachtert**
These changes add a new setting 'groupsFieldName'. If it's set and the OpenID Connect server adds an array with that name containing group names in the ID token, then those groups are added as groups within Jenkins. This allows group based authorization in Jenkins.

Any comments welcome, this is +/- the first time I ever changed a Jenkins plugin.

### CODE REVIEW

**Comment by ashb** [comment]
Thanks for this - it worked for me and meant my groups showed up in Jenkins from Azure AD/Azure OpenID Connect.

(They show up as UUIDs in Jenkins, but that's what Azure OIDC returns them as.)

**Comment by jordanjennings** [comment]
This is extremely useful functionality. Would love to see this get merged! Is there any holdup here?

**Comment by jordanjennings** [comment]
FYI - I built this plugin using `mvn package` (is that the right target?) and installed using the web UI on my Jenkins instance which is running the latest LTS (2.89.1) and it totally broke everything. Unfortunately I didn't think to copy the stack trace in my hurry to roll it back. But seems this needs some updating to be compatible with the latest version of Jenkins or something. If I get more time I'll stand up a fresh Jenkins and try it out again. Or if there's some other way I'm supposed to create the .hpi please let me know.

**Comment by ashb** [comment]
@jordanjennings `mvn package` or `mvn install`, something like that, is how I built it in August and it worked fine for me then -- I don't have time to test again right now, sorry

**Comment by DriesVerachtert** [comment]
Note you'll have to gunzip it first (github doesn't allow .hpi attachments): [oic-auth.hpi.gz](https://github.com/jenkinsci/oic-auth-plugin/files/1547435/oic-auth.hpi.gz)
That's the modified plugin that I'm using at the moment. Of course it's better that you build your own.

@jordanjennings, you have to make sure that at least some group or user still has admin access to your jenkins instance before you enable this plugin. It's fairly easy to lock out yourself if you use this plugin for role-based authorization without defining some admin group first, and that group of course has to exist in the OIDC.

**Comment by neith00** [comment]
It would be great to merge this PR!

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 1]
```diff
@@ -172,6 +187,54 @@ public String getLoginUrl() {
...
+		private GrantedAuthority[] grantedAuthorities;
+
+		OicUserDetails(String userName, GrantedAuthority[] grantedAuthorities) {
+			this.userName = userName;
+			this.grantedAuthorities = grantedAuthorities;
+		}
+		
+		@Override
+		public GrantedAuthority[] getAuthorities() {
+			LOGGER.info("OicUserDetails.getAuthorities called, returning " + grantedAuthorities.length);
```
Comment (mjmbischoff) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: should probably be logging on lower level

**Pull request merged**